### PR TITLE
[WIP] pytest

### PR DESCRIPTION
**Pull request by shimizukawa**
**THIS PR #3300 IS DEPRECATED**

Using pytest test runner instead of nose.
I have still some todos.

[ADD]

Based on @tk0miya 's proposal, I will create the following PR in stages:

1. PR for pytest migration into stable branch -> #3302 
2. refactoring for pytest -> #3309
3. Add sphinx.test that will replace the sphinx-testing package.  -> not yet
4. speed-up tests -> not yet

### CODE REVIEW

File: tests/test_build.py [hunk 1]
```diff
@@ -64,36 +50,40 @@ def test_build_all():
...
+        app = make_app(buildername, srcdir=nonascii_srcdir)
+        # if listdir is not empty, we can use built cache
+        if not app.outdir.listdir():
+            app.build()
 
 
-@with_tempdir
-def test_master_doc_not_found(tmpdir):
+def test_master_doc_not_found(tmpdir, make_app):
+    tmpdir = path(tmpdir)
```
Comment (tk0miya) [inline tests/test_build.py]: How about making `tempdir` fixture instead `tmpdir`? Wrapping everytime are very boring for me.

File: tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            app_.build()
+    yield app_
+
+    if test_params['shared_result']:
+        shared_result.store(test_params['shared_result'], app_)
+
+
+@pytest.fixture(scope='function')
+def status(app):
+    return app.test_status
```
Comment (tk0miya) [inline tests/conftest.py]: Why are you using `app.test_status`? I prefer to use `app._status` instead.

File: tests/test_build.py (hunk 1 above)
Comment (shimizukawa) [reply tests/test_build.py]: I agree for "boring". I'll try it.

**Comment by tk0miya** [comment]
I know this is still WIP. But, it would be nice if you can separate refactoring testcases and pytest migration.
It seems this PR contains both changes. So it is hard to review this for me.

In addition, I prefer to merge this into stable branch.  If we're going to merge this into master, to merge stable branch into master will be much harder.  It prevents our maintenance works.

**Comment by tk0miya** [comment]
One more thing, we already have https://github.com/sphinx-doc/sphinx-testing as a testing framework.
So I prefer to upgrade it, and Sphinx uses it.
What do you think?

File: tests/test_build.py (hunk 1 above)
Comment (shimizukawa) [inline tests/test_build.py]: +1

**Comment by shimizukawa** [comment]
Thanks for your comments @tk0miya !
Based on your proposals, I updated description of this PR.

**Comment by shimizukawa** [comment]
Now this PR is closed but please DON'T REMOVE `pytest` branch for a while.

**Pull request closed without merging**